### ISSUE TITLE
Support for retrofit2.mock.Calls.failure() to accept exceptions types other than IOException

### ISSUE THREAD

**Issue by cpeppas** [square/retrofit#2693]
- [x] Feature Request. 

I have implemented unit tests for my api using the MockRetrofit and BehaviorDelegate

So far I am able to assert success 200, Http error (400,500) and Network (IoException) error cases.

I would like also to be able to assert error cases such as NullPointerException (mainly parsing issues that can be possible thrown) or any other Exception other than IOException

Example

```
    private class MockApi implements MyApi {
        final BehaviorDelegate<MyApi> delegate;

        MockApi(BehaviorDelegate<MyApi> delegate) {
            this.delegate = delegate;
        }

        @Override public Single<MyResponse> retrieveStuff(String param) {
            return delegate.returning(Calls.failure(new MyException())).retrieveStuff(param);
        }
    }
```

**Comment by JakeWharton** [comment]
PR welcome!

**Comment by naturalwarren** [comment]
I'd love to help here if there's any interest. @cpeppas Is this something you're already working on?

**Comment by cpeppas** [comment]
Thanks @naturalwarren I haven't started anything yet so feel free 👍

**Comment by JakeWharton** [comment]
Merged as ce787ed5.
